datafusion/sql/src/select.rs

```diff
@@ -276,33 +277,59 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
         Ok(plan)
     }
 
-    // Try converting Expr::Unnest to LogicalPlan::Unnest if possible, otherwise do the final projection
+    /// Try converting Expr(Unnest(Expr)) to Projection/Unnest/Projection
     pub(super) fn try_process_unnest(
         &self,
         input: LogicalPlan,
         select_exprs: Vec<Expr>,
     ) -> Result<LogicalPlan> {
         let mut unnest_columns = vec![];
-        // Map unnest expressions to their argument
-        let projection_exprs = select_exprs
+        let mut inner_projection_exprs = vec![];
+
+        let outer_projection_exprs = select_exprs
             .into_iter()
             .map(|expr| {
-                if let Expr::Unnest(Unnest { ref exprs }) = expr {
+                let Transformed {
+                    data: transformed_expr,
+                    transformed,
+                    tnr: _,
+                } = expr.transform_up_mut(&mut |expr: Expr| {
```

Comment:
Why the logic of unnest is split out of others two?
is doing if else directly without transformed more readable?
```rust
if unnest
else if column
else {
}
```

Reply:
@jayzhan211 The transform is essential for this feature. We need to rewrite the expression in bottom-up recursion such that we can extract the child and parent expression of the unnest function. The reason to split out the Column variant is that: we need to retain the `relation` field

Reply:
```rust
                let column_name = expr.display_name()?;
                if let Expr::Unnest(Unnest { ref exprs }) = expr {
                    unnest_columns.push(column_name.clone());
                    // Add alias for the argument expression, to avoid naming conflicts with other expressions
                    // in the select list. For example: `select unnest(col1), col1 from t`.
                    inner_projection_exprs
                        .push(exprs[0].clone().alias(column_name.clone()));
                    Ok(Expr::Column(Column::from_name(
                        column_name,
                    )))
                } else if let Expr::Column(_) = expr {
                    inner_projection_exprs.push(expr.clone());
                    Ok(expr)
                } else {
                    // We need to evaluate the expr in the inner projection,
                    // outer projection just select its name
                    inner_projection_exprs.push(expr);
                    Ok(Expr::Column(Column::from_name(column_name)))
                }
```

I'm not sure where the recursion of the rewrite happens, is there any example? I pass the test with the code above.

Reply:
Ok, I got it. First time seeing `transform_up_mut`, not familiar with it.